lib/Backend/arm64/LowerMD.cpp

```diff
@@ -1353,89 +1341,95 @@ LowererMD::LowerEntryInstr(IR::EntryInstr * entryInstr)
-    Callee-Saved Int Registers (r19-r28)
-    Local Variables
-    lr - link Register (r30)
-    fp - Frame Pointer (r29)
-    alloca area
-    */
-
     // Zero-initialize dedicated arguments slot
     if (hasCalls)
     {
         //R17 acts a dummy zero register which we push to arguments slot
         //mov r17, 0
         Assert(r17Opnd == nullptr);
         r17Opnd = IR::RegOpnd::New(nullptr, SCRATCH_REG, TyMachReg, this->m_func);
-        IR::Instr * instrMov = IR::Instr::New(Js::OpCode::MOV, r17Opnd, IR::IntConstOpnd::New(0, TyMachReg, this->m_func), this->m_func);
+        IR::Instr * instrMov = IR::Instr::New(Js::OpCode::MOVZ, r17Opnd, IR::IntConstOpnd::New(0, TyMachReg, this->m_func), this->m_func);
         insertInstr->InsertBefore(instrMov);
         IR::LabelInstr *prologStartLabel = IR::LabelInstr::New(Js::OpCode::Label, this->m_func);
         insertInstr->InsertBefore(prologStartLabel);
         this->m_func->m_unwindInfo.SetPrologStartLabel(prologStartLabel->m_id);
     }
 
+    int32 regPushedSize = 0;
+
     if (!paramRegs.IsEmpty())
     {
         for (int i = LAST_INT_ARG_REG; i >= FIRST_INT_ARG_REG; --i)
         {
             RegNum reg = (RegNum)(i);
             if (paramRegs.Test(RegEncode[reg]))
             {
+                regPushedSize += MachRegInt;
                 IR::Instr * instrStore = IR::Instr::New(Js::OpCode::STR, this->m_func);
-                instrStore->SetDst(IR::IndirOpnd::New(IR::RegOpnd::New(nullptr, RegSP, TyMachReg, this->m_func), (int32)0, TyMachReg, this->m_func));
+                instrStore->SetDst(IR::IndirOpnd::New(IR::RegOpnd::New(nullptr, RegSP, TyMachReg, this->m_func), -regPushedSize, TyMachReg, this->m_func));
                 instrStore->SetSrc1(IR::RegOpnd::New(reg, TyMachReg, this->m_func));
                 insertInstr->InsertBefore(instrStore);
             }
         }
-
-        /*
-        // Generate PUSH {r0-r3}
-        IR::Instr * instrPush = IR::Instr::New(Js::OpCode::STP, this->m_func);
-        instrPush->SetDst(IR::IndirOpnd::New(IR::RegOpnd::New(nullptr, RegSP, TyMachReg, this->m_func), (int32)0, TyMachReg, this->m_func));
-        instrPush->SetSrc1(IR::RegBVOpnd::New(paramRegs, TyMachReg, this->m_func));
-        insertInstr->InsertBefore(instrPush);
-        */
     }
 
     // Setup Frame pointer
     if (hasCalls)
     {
-        BVUnit frameRegs;
-        frameRegs.Set(RegEncode[RegFP]);
-        frameRegs.Set(RegEncode[RegLR]);
-
-        // Generate PUSH {r11,lr}
-        IR::Instr * instrPush = IR::Instr::New(Js::OpCode::STP, this->m_func);
-        instrPush->SetDst(IR::IndirOpnd::New(IR::RegOpnd::New(nullptr, RegSP, TyMachReg, this->m_func), (int32)0, TyMachReg, this->m_func));
-        instrPush->SetSrc1(IR::RegBVOpnd::New(frameRegs, TyMachReg, this->m_func));
-        insertInstr->InsertBefore(instrPush);
+        // Generate STP {fp,lr}
+        IR::Instr * instrStp = IR::Instr::New(Js::OpCode::STP_PRE, this->m_func);
```

Comment:
The limit of pre-indexed STP is -512, so unless you can guarantee that, you will need to add fallback logic to do a manual stack allocation.

Reply:
We are only pushing R0-R7, fp, lr so its 10*8 which is < 512. Added an assert for now.

---
In reply to: [143538688](https://github.com/Microsoft/ChakraCore/pull/3903#discussion_r143538688) [](ancestors = 143538688)

Reply:
As long as you can guarantee it, you're good. Just wanting to be careful.